### ISSUE TITLE
Issue in raising exception for CSRR instruction in RISCV 

### ISSUE THREAD

**Issue by spidugu444** [hubbsvtgc/LearnRISC-V#5]
Hi, I am working on verifying access modes for RISCV CSRs, I am trying to verify Machine Mode Read only CSRs in lower privileged mode i.e on Supervisor Mode.

As per RISCV privileged Specification if we try to access Machine mode CSRs in lower privileged modes it will raise exception.

When i tried back-to-back read on MIMPID CSR in supervisor Mode it is expexted to raise exception for all access in supervisor mode but in my case it raises exception for first CSRR instruction for second and last CSRR instruction it didn't raised exception.
could anyone please help me in resolving this issue.

please find the test code [here](https://gist.github.com/spidugu444/d629dfa878b2cc1cbdc8ff96bf4529bb#file-riscv_mimpid_test-s) and let me know if any info needed.

Thanks

**Comment by hubbsvtgc** [comment]
I find it hard to read the code, as csr register number is used instead of csr name, few observations though. The supervisor_exception_handler is mapped to mtvec meaning its mapped to machine mode trap handling.  Unable to read what the code does inside csr_fail.  It seems like the first exception would take the execution to the exception handler (in M mode) and the next iterations execute in M-mode and hence no more exception.

**Comment by spidugu444** [comment]
@hubbsvtgc, I added csr_pas and csr_fail snippet, please find it in above code link.
In above code I am switching processor to supervisor mode then i am trying to access Machine mode CSR, then only for first exception it is raising exception for  remaining csr access it is not raising exceptions.

**Comment by hubbsvtgc** [comment]
@spidugu444  can you check if the exception handler after first exception, returns to supervisor mode? or it returns in machine mode itself ?

**Comment by spidugu444** [comment]
@hubbsvtgc, After first exception still the core is in supervisor mode i checked it by reading MSTATUS CSR which is showing 0x800.

**Comment by spidugu444** [comment]
@hubbsvtgc, After raising first exception MSTATUS bit has been modified to Machine Mode even after modifying code so that every time core will be in supervisor mode every time, but even though it is not raising exception for second and last CSRR instruction.

**Comment by hubbsvtgc** [comment]
@spidugu444 "  _after modifying code so that every time core will be in supervisor mode every time_  " , I don't find this happening in code, its hard with only **partial code** . The closest reference to help would be https://github.com/hubbsvtgc/LearnRISC-V/blob/dev/apps/m2u/user-machine-switch.S.  To change the privilege level, MPP should be the intended privilege level along with xret instruction. 

If you watch [here](https://www.youtube.com/watch?v=G897K_V6dQ0) machine to user mode switch would happen only after mret execution in trap handler. I don't find corresponding xret instruction in the code. Jump instruction wont switch the mode.

**Comment by spidugu444** [comment]
Hi @hubbsvtgc, Thanks for your reference. after raising first exception MSTATUS MPP  bit is moving to machine mode and i used below code to switch into supervisor mode but still it is not working for me.

supervisor_exception_handler:
	## Switching MPP bit to 01(supervisor mode)
	csrr x3, 0x300
	li x29, 0b1
        slli x29, x29, 11
        addi x28, x29, 0
 
         # Write the modified MSTATUS value back to the CSR
        csrw 0x300, x28

        csrr x30, 0x341	    ## Reading MEPC CSR which holds exception origin Address		
        csrr x31, 0x342     ## Reading MCAUSE CSR which holds the cause of exception
        li x2 ,2
        beq x31, x2, next_iter1   ## Checking is exception is expected exception or not
        j csr_fail

next_iter1:
	csrw 0x342, 0		## Reseting MCAUSE value to 0 before handling new exception
	beq x30, x0, csr_fail

	addi x7, x30, 12
        jr x7			## Jump to MEPC + 12 Address location  skipping 3 instructions


@hubbsvtgc, could you please help me in resolving this issue.

**Comment by hubbsvtgc** [comment]
@spidugu444 I don't see this snippet [here](https://gist.github.com/spidugu444/d629dfa878b2cc1cbdc8ff96bf4529bb#file-riscv_mimpid_test-s) .  **I would close this issue now as its hard to follow your code and snippet here **. Try to break it down and try to resolve it step by step. Also as I mentioned, unless [X]ret is executed, the mode switch wont happen. First you need to upload your local changes to the repo to be visible. Cant work on broken pieces of code which are to be glued together.
